cluster-autoscaler/core/scaledown/actuation/delete_in_batch_test.go

```diff
@@ -0,0 +1,216 @@
+		failedDeletion int
+		addNgToBucket  bool
+	}{
+		{
+			name:          "Remove NodeGroup that is not present in bucket",
+			err:           true,
+			addNgToBucket: false,
+		},
+		{
+			name:           "Regular successful remove",
+			err:            false,
+			numNodes:       5,
+			failedDeletion: 0,
+			addNgToBucket:  true,
+		},
+		{
+			name:           "Unsuccessful remove",
+			err:            true,
+			numNodes:       5,
+			failedDeletion: 1,
+			addNgToBucket:  true,
+		},
+	}
+	for _, test := range testCases {
+
+		fakeClient := &fake.Clientset{}
+		fakeLogRecorder, _ := clusterstate_utils.NewStatusMapRecorder(fakeClient, "kube-system", kube_record.NewFakeRecorder(5), false, "my-cool-configmap")
+
+		failedNodeDeletion := make(map[string]bool)
+		deletedNodes := make(chan string, 10)
+		nodDeletedNodes := make(chan string, 10)
+		// Hook node deletion at the level of cloud provider, to gather which nodes were deleted, and to fail the deletion for
+		// certain nodes to simulate errors.
+		provider := testprovider.NewTestCloudProvider(nil, func(nodeGroup string, node string) error {
+			if failedNodeDeletion[node] {
+				nodDeletedNodes <- node
+				return fmt.Errorf("SIMULATED ERROR: won't remove node")
+			}
+			deletedNodes <- node
+			return nil
+		})
+
+		ctx, err := NewScaleTestAutoscalingContext(config.AutoscalingOptions{}, nil, nil, provider, nil, nil)
+		clusterStateRegistry := clusterstate.NewClusterStateRegistry(provider, clusterstate.ClusterStateRegistryConfig{}, fakeLogRecorder, NewBackoff())
+		if err != nil {
+			t.Fatalf("Couldn't set up autoscaling context: %v", err)
+		}
+
+		ng := "ng"
+		provider.AddNodeGroup(ng, 1, 10, test.numNodes)
+
+		d := NodeBatcherDeleter{
+			ctx:                   &ctx,
+			clusterState:          clusterStateRegistry,
+			nodeDeletionTracker:   deletiontracker.NewNodeDeletionTracker(1 * time.Minute),
+			deletionsPerNodeGroup: make(map[string][]*apiv1.Node),
+			drainedNodeDeletions:  make(map[string]bool),
+		}
+		nodes := generateNodes(test.numNodes, ng)
+		failedDeletion := test.failedDeletion
+		for _, node := range nodes {
+			if failedDeletion > 0 {
+				failedNodeDeletion[node.Name] = true
+				failedDeletion -= 1
+			}
+			provider.AddNode(ng, node)
+		}
+		if test.addNgToBucket {
+			for _, node := range nodes {
+				node.Spec.Taints = append(node.Spec.Taints, apiv1.Taint{
+					Key:    deletetaint.ToBeDeletedTaint,
+					Effect: apiv1.TaintEffectNoSchedule,
+				})
+				_, _, err := d.addNodeToBucket(node, true)
+				if err != nil {
+					t.Errorf("%s: addNodeToBucket return error %q when addidng node %v", test.name, err, node)
+				}
+			}
+		}
+
+		err = d.remove(ng)
+		if test.err {
+			if err == nil {
+				t.Errorf("%s: remove() should return error, but return nil", test.name)
+			}
+			return
+		}
+		if err != nil {
+			t.Errorf("%s: remove() return error, but shouldn't", test.name)
+		}
+		gotDeletedNodes := []string{}
+		for i := 0; i < test.numNodes; i++ {
```

Comment:
Right, the Unsuccessful remove case hasn't run because of return statement, now it fixed.
If we have failed deletion the following nodes are not deleted, so I check this deleted nodes only in success case and otherwise I check at least one failure.